Fix process_array_expr to take into account pointer offset

tautschnig opened the pull request:
Includes code cleanup to remove an unnecessary use of process_array_expr.

tautschnig committed:
Process array_equal the same way as array_{replace,copy}

array_equal was handled as a special snowflake in both goto-program conversion
and symbolic execution. This simplifies the code and removes a sharing-unaware
pass that ran over all expressions. This commit is only refactoring, no
behavioural changes expected.

This commit also moves a regression test from cbmc-from-CVS to the cbmc folder,
and extends it to cover array_replace.

tautschnig committed:
Fix process_array_expr to take into account pointer offset

peterschrammel reviewed (commented):
Nice piece of work. 
Is there a test case for array_replace? I don't see how it is called in your test case.

tautschnig commented:
> Is there a test case for array_replace? I don't see how it is called in your test case.

It is called from (our models of) the C library functions, including `memcpy`. Is that sufficient?

peterschrammel reviewed (approved):
Found it. I searched in the wrong file.

tautschnig committed:
Byte offset to array index translation must use array type

When dereferencing a void* pointer, the dereferenced type would always be
considered compatible with the underlying object type. Nevertheless they are not
actually the same. If the object was an array and the byte offset non-zero, the
use of void* would yield the wrong index. The correct way to compute the index
is using the actual (array) object type.

Fixes: #1857

chrisr-diffblue commented on src/goto-symex/symex_clean_expr.cpp:
```diff
@@ -137,10 +60,57 @@ void goto_symext::process_array_expr(exprt &expr)
+      }
+
+      // given an array type T[S] and a byte offset o, build a new type
+      // T[S-(o/sizeof(T))]
```
I may have been particularly slow reading this comment (apologies if so) - but it wasn't immediately obvious to me that this type conversion (particularly the `o/sizeof(T)`) is to do the conversion from byte offset -> index. Maybe it's worth expanding the comment very slightly?

martin-cs commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -221,8 +221,6 @@ void ansi_c_internal_additions(std::string &code)
     // arrays
-    // NOLINTNEXTLINE(whitespace/line_length)
-    "__CPROVER_bool __CPROVER_array_equal(const void *array1, const void *array2);\n"
     // overwrite all of *dest (possibly using nondet values), even
```
I like the idea of removing special case passes, particularly ones that are sharing unaware BUT it seems like array_copy only captures some of the functionality of array_equals.  With array_equals I can build two arrays in different ways and then test their equality.

martin-cs commented on src/goto-symex/symex_clean_expr.cpp:
```diff
@@ -94,41 +30,28 @@ void goto_symext::process_array_expr(exprt &expr)
-    expr.swap(tmp);
-    process_array_expr(expr);
-  }
   else if(expr.id()==ID_symbol &&
```
There seem to be quite a few cases that have been refactored here that, to my eyes, are not covered by the test case added.

martin-cs reviewed (changes requested):
Am concerned about the change in functionality but otherwise OK.

tautschnig commented:
@martin-cs Thank you for getting me back on track, I'm not sure what I was doing when crafting ad728880fa55. I'll rework.

martin-cs commented:
@tautschnig As I said about another commit -- about half of it is "hell yes we should have done that years ago"...

tautschnig commented:
I have reworked all the array_equal stuff and now also included a specific regression test (rather than the more implicit ones before).

tautschnig commented:
@martin-cs Would you mind taking another look?

martin-cs reviewed (approved):
Looks much better.  I would still prefer some more tests for the second commit but I won't block the merge because of it.

tautschnig commented:
Just to note that the request for more tests is not being ignored: I'm trying to source them, but please bear with me as that might take some time.

martin-cs commented:
Sure; it wasn't supposed to be a blocking request more a "I'd feel more confident about this if...".
